File: docs/features/consumption/_snippets/cpp/stream_multi_source.cpp
#include <iostream>
#include <string>
#include <vector>

#include "ouster/core/frame_set.h"
#include "ouster/core/open_source.h"
#include "ouster/sensor/sensor_frame_set_source.h"

using namespace ouster::sdk;

namespace {

void iterate_source(core::FrameSetSource& source, std::size_t limit = 5) {
    std::size_t count = 0;
    // clang-format off
    // [doc-stag-multi-sensorframesetsource-loop]
    for (const auto& frame_set : source) {
        for (std::size_t idx = 0; idx < frame_set.size();++idx) {
            const auto& frame = frame_set[idx];
            if (frame) {
                std::cout << "frame: " << count
                          << " id=" << frame->frame_id << "\n";
                // [doc-etag-multi-sensorframesetsource-loop]
                // clang-format on
            } else {
                continue;
            }
        }
        if (++count >= limit) {
            break;
        }
    }
}

}  // namespace

void stream_multi_sensor(const std::vector<std::string>& hostnames, std::size_t limit = 5) {
    // [doc-stag-multi-sensorframesetsource]
    // Replace with actual hostnames
    // std::vector<std::string> hostnames{"os-xxxx.local", "os-yyyy.local"};
    sensor::SensorFrameSetSource source(hostnames);
    // [doc-etag-multi-sensorframesetsource]
    iterate_source(source, limit);
}

void stream_multi_open_source(const std::vector<std::string>& hostnames, std::size_t limit = 5) {
    // clang-format off
    // [doc-stag-multi-opensource-sensor]
    // Replace with actual hostnames
    // auto hostnames = {"os-xxxx.local", "os-yyyy.local"};
    auto source = open_source(hostnames);
    //! [doc-etag-multi-opensource-sensor]
    std::size_t count = 0;
    for (const auto& frame_set : source) {
        for (std::size_t idx = 0; idx < frame_set.size();++idx) {
            const auto& frame = frame_set[idx];
            if (frame) {
                std::cout << "frame: " << count
                          << " id=" << frame->frame_id << "\n";
            } }
        if (++count >= limit) { break; } }
    // clang-format on
}

void stream_multi_open_source(const std::vector<std::string>& hostnames,
                              const std::vector<std::string>& metadata_paths,
                              std::size_t limit = 5) {
    auto source =
        open_source(hostnames, [&](FrameSetSourceOptions& opt) { opt.meta = metadata_paths; });
    iterate_source(source, limit);
}

void replay_multi_recording(const std::vector<std::string>& paths, std::size_t limit = 5) {
    // clang-format off
    // [doc-stag-multi-opensource-file]
    // auto paths = {"capture.pcap", "capture2.pcap"};
    auto source = open_source(paths);

    std::size_t count = 0;
    for (const auto& frame_set : source) {
        for (std::size_t idx = 0; idx < frame_set.size();++idx) {
            const auto& frame = frame_set[idx];
            if (frame) {
                std::cout << "frame: " << count
                          << " id=" << frame->frame_id << "\n";
            } else {
                continue;
            }
        }
        if (++count >= limit) {
            break;
        }
    }
    // [doc-etag-multi-opensource-file]
    // clang-format on
}

File: docs/features/consumption/_snippets/python/stream_multi_source.py
"""Doc snippets for multi-sensor streams with SensorFrameSetSource/open_source."""
from typing import Iterable, Optional, Sequence
from ouster import sdk
from ouster.sdk import sensor
from ouster.sdk.core import LidarFrame


def _print_frame(count: int, frame_set: Iterable[Optional[LidarFrame]]) -> None:
    for idx, frame in enumerate(frame_set):
        if frame:
            print(f"frame: {count} frame_id={frame.frame_id}")


def stream_multi_sensor(hosts: Sequence[str], limit: int = 5) -> None:
    """Connect to multiple live sensors with SensorFrameSetSource."""
    # [doc-stag-multi-sensorframesetsource]
    # Replace with actual hostnames
    # hosts = ["os-xxxx.local", "os-yyyy.local"]
    source = sensor.SensorFrameSetSource(list(hosts))
    # [doc-etag-multi-sensorframesetsource]
    # [doc-stag-multi-sensorframesetsource-loop]
    with source:
        for count, frame_set in enumerate(source):
            for idx, _ in enumerate(frame_set):
                frame = frame_set[idx]
                if frame:
                    print(f"frame:{count} frame_id={frame.frame_id}")
        # [doc-etag-multi-sensorframesetsource-loop]
            if count + 1 >= limit:
                break


def stream_multi_open_source(hosts: Sequence[str], limit: int = 5) -> None:
    """Same as stream_multi_sensor but via open_source (collates by default)."""
    # [doc-stag-multi-opensource-sensor]
    # Replace with actual hostnames
    # hostnames = ["os-xxxx.local", "os-yyyy.local"]
    source = sdk.open_source(list(hosts), collate=True)
    with source:
        for count, frame_set in enumerate(source):
            for idx, _ in enumerate(frame_set):
                frame = frame_set[idx]
                if frame:
                    print(f"frame:{count}" f"frame_id={frame.frame_id}")

            if count + 1 >= limit:
                break
    # [doc-etag-multi-opensource-sensor]


def replay_multi_recording(paths: Sequence[str], limit: int = 5) -> None:
    """Replay multi-sensor captures (OSF/PCAP + metadata) with open_source."""
    # [doc-stag-multi-opensource-file]
    # paths = ["capture.osf", "capture2.osf"]
    source = sdk.open_source(list(paths), collate=True)
    # [doc-etag-multi-opensource-file]
    with source:
        for count, frame_set in enumerate(source):
            _print_frame(count, frame_set)
            if count + 1 >= limit:
                break


if __name__ == "__main__":
    pass
